File: scripts/auto-capture.js
/**
 * Auto-capture ForiFlow screenshots for the GitHub README.
 *
 *   npm run capture
 *   node scripts/auto-capture.js
 */

const fs = require("fs");
const http = require("http");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const OUT = path.join(ROOT, "docs", "screenshots");
const FRONTEND = "http://localhost:3000";
const BACKEND = "http://localhost:8000";

const FORM = [
  ["applicant_name", "Ali Khan"],
  ["business_name", "Khan Traders"],
  ["loan_amount_pkr", "500000"],
  ["tenure_months", "12"],
  ["monthly_digital_payments", "150000"],
  ["payment_history_score", "0.95"],
  ["inventory_turnover", "4.5"],
  ["order_consistency", "1.0"],
  ["existing_debt_pkr", "100000"],
  ["cash_flow_proxy", "1.0"],
  ["years_in_operation", "5"],
  ["num_employees", "8"],
];

function httpOk(url) {
  return new Promise((resolve) => {
    const req = http.get(url, (res) => {
      res.resume();
      resolve(res.statusCode >= 200 && res.statusCode < 500);
    });
    req.setTimeout(3000, () => {
      req.destroy();
      resolve(false);
    });
    req.on("error", () => resolve(false));
  });
}

function printStartCommands() {
  console.log("");
  console.log("Servers are not running. Start them in two terminals, then re-run npm run capture:");
  console.log("");
  console.log("  Terminal 1: cd backend && python -m uvicorn main:app --reload --port 8000");
  console.log("  Terminal 2: cd frontend && npm run dev -- --port 3000");
  console.log("");
}

async function clickNav(page, label) {
  await page.getByRole("link", { name: label }).first().click();
  await page.waitForTimeout(500);
}

async function main() {
  const backendUp = await httpOk(`${BACKEND}/health`);
  const frontendUp = await httpOk(`${FRONTEND}/`);

  if (!backendUp || !frontendUp) {
    printStartCommands();
    console.log("Waiting 10 seconds...");
    await new Promise((r) => setTimeout(r, 10000));
    const backend2 = await httpOk(`${BACKEND}/health`);
    const frontend2 = await httpOk(`${FRONTEND}/`);
    if (!backend2 || !frontend2) {
      printStartCommands();
      process.exit(1);
    }
  }

  let chromium;
  try {
    ({ chromium } = require("@playwright/test"));
  } catch {
    try {
      ({ chromium } = require("playwright"));
    } catch {
      console.error("Install Playwright first: npm install && npx playwright install chromium");
      process.exit(1);
    }
  }

  fs.mkdirSync(OUT, { recursive: true });

  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage({
    viewport: { width: 1920, height: 1080 },
    colorScheme: "light",
  });

  await page.goto(`${FRONTEND}/`, { waitUntil: "networkidle" });
  await page.getByText(/API online|API Online|Dashboard/i).first().waitFor({ timeout: 20000 });
  await page.screenshot({ path: path.join(OUT, "01-dashboard.png"), fullPage: true });

  await clickNav(page, "Credit Scoring");
  await page.locator("#applicant_name").waitFor({ timeout: 15000 });
  await page.screenshot({ path: path.join(OUT, "02-scoring-form.png"), fullPage: true });

  for (const [name, value] of FORM) {
    const input = page.locator(`#${name}`);
    await input.fill("");
    await input.fill(value);
  }
  await page.getByRole("button", { name: /Score application/i }).click();
  await page.getByText(/Approved|Rejected|Manual Review/).first().waitFor({ timeout: 60000 });
  await page.waitForTimeout(800);
  await page.screenshot({ path: path.join(OUT, "03-score-result.png"), fullPage: true });

  const shapBtn = page.getByRole("button", { name: /View SHAP/i }).first();
  if (await shapBtn.count()) {
    await shapBtn.click();
  } else {
    await clickNav(page, "SHAP Reports");
  }
  await page.locator("svg").first().waitFor({ timeout: 20000 });
  await page.waitForTimeout(600);
  await page.screenshot({ path: path.join(OUT, "04-shap-chart.png"), fullPage: true });

  await clickNav(page, "EWS Alerts");
  await page.locator("table, h2").first().waitFor({ timeout: 15000 });
  await page.screenshot({ path: path.join(OUT, "05-ews-alerts.png"), fullPage: true });

  await clickNav(page, "Applications");
  await page.locator("table tbody tr, table").first().waitFor({ timeout: 20000 });
  await page.screenshot({ path: path.join(OUT, "06-applications.png"), fullPage: true });

  await page.goto(`${BACKEND}/docs`, { waitUntil: "networkidle" });
  await page.locator(".swagger-ui, #swagger-ui").first().waitFor({ timeout: 20000 });
  await page.screenshot({ path: path.join(OUT, "07-swagger.png"), fullPage: true });

  await browser.close();
  console.log("✅ Screenshots saved to docs/screenshots/");
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});

File: scripts/auto-capture.py
"""Python fallback for ForiFlow README screenshots.

    pip install playwright
    playwright install chromium
    python scripts/auto-capture.py
"""

from __future__ import annotations

import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "docs" / "screenshots"
FRONTEND = "http://localhost:3000"
BACKEND = "http://localhost:8000"

FORM = [
    ("applicant_name", "Ali Khan"),
    ("business_name", "Khan Traders"),
    ("loan_amount_pkr", "500000"),
    ("tenure_months", "12"),
    ("monthly_digital_payments", "150000"),
    ("payment_history_score", "0.95"),
    ("inventory_turnover", "4.5"),
    ("order_consistency", "1.0"),
    ("existing_debt_pkr", "100000"),
    ("cash_flow_proxy", "1.0"),
    ("years_in_operation", "5"),
    ("num_employees", "8"),
]


def http_ok(url: str) -> bool:
    try:
        with urllib.request.urlopen(url, timeout=3) as response:
            return 200 <= response.status < 500
    except (urllib.error.URLError, TimeoutError, OSError):
        return False


def print_start_commands() -> None:
    print()
    print("Servers are not running. Start them in two terminals, then re-run:")
    print()
    print("  Terminal 1: cd backend && python -m uvicorn main:app --reload --port 8000")
    print("  Terminal 2: cd frontend && npm run dev -- --port 3000")
    print()


def click_nav(page, label: str) -> None:
    page.get_by_role("link", name=label).first.click()
    page.wait_for_timeout(500)


def main() -> int:
    if not (http_ok(f"{BACKEND}/health") and http_ok(f"{FRONTEND}/")):
        print_start_commands()
        print("Waiting 10 seconds...")
        time.sleep(10)
        if not (http_ok(f"{BACKEND}/health") and http_ok(f"{FRONTEND}/")):
            print_start_commands()
            return 1

    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        print("Install Playwright first: pip install playwright && playwright install chromium")
        return 1

    OUT.mkdir(parents=True, exist_ok=True)

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=True)
        page = browser.new_page(
            viewport={"width": 1920, "height": 1080},
            color_scheme="light",
        )

        page.goto(f"{FRONTEND}/", wait_until="networkidle")
        page.get_by_text("API online").or_(page.get_by_text("Dashboard")).first.wait_for(
            timeout=20000
        )
        page.screenshot(path=str(OUT / "01-dashboard.png"), full_page=True)

        click_nav(page, "Credit Scoring")
        page.locator("#applicant_name").wait_for(timeout=15000)
        page.screenshot(path=str(OUT / "02-scoring-form.png"), full_page=True)

        for name, value in FORM:
            field = page.locator(f"#{name}")
            field.fill("")
            field.fill(value)
        page.get_by_role("button", name="Score application").click()
        page.get_by_text("Approved").or_(page.get_by_text("Rejected")).or_(
            page.get_by_text("Manual Review")
        ).first.wait_for(timeout=60000)
        page.wait_for_timeout(800)
        page.screenshot(path=str(OUT / "03-score-result.png"), full_page=True)

        button = page.get_by_role("button", name="View SHAP")
        if button.count():
            button.first.click()
        else:
            click_nav(page, "SHAP Reports")
        page.locator("svg").first.wait_for(timeout=20000)
        page.wait_for_timeout(600)
        page.screenshot(path=str(OUT / "04-shap-chart.png"), full_page=True)

        click_nav(page, "EWS Alerts")
        page.locator("table, h2").first.wait_for(timeout=15000)
        page.screenshot(path=str(OUT / "05-ews-alerts.png"), full_page=True)

        click_nav(page, "Applications")
        page.locator("table tbody tr, table").first.wait_for(timeout=20000)
        page.screenshot(path=str(OUT / "06-applications.png"), full_page=True)

        page.goto(f"{BACKEND}/docs", wait_until="networkidle")
        page.locator(".swagger-ui, #swagger-ui").first.wait_for(timeout=20000)
        page.screenshot(path=str(OUT / "07-swagger.png"), full_page=True)

        browser.close()

    print("✅ Screenshots saved to docs/screenshots/")
    return 0


if __name__ == "__main__":
    sys.exit(main())
